File: 2021/3/puzz2.py
with open('input', 'r') as f:
   nums = f.read().split('\n')[:-1]

def extract_num(nums, criteria):
   i = 0
   while len(nums) > 1:
      ones = 0
      zeros = 0
      for num in nums:
         if num[i] == '1':
            ones += 1
         else:
            zeros += 1

      if (zeros > ones and criteria == 'most') or (zeros <= ones and criteria == 'least'):
         nums = [num for num in filter(lambda num: num[i] == '0', nums)]
      else:
         nums = [num for num in filter(lambda num: num[i] == '1', nums)]
      i += 1

   return nums[0]

oxy_num = extract_num(nums, 'most')
co2_num = extract_num(nums, 'least')
print(oxy_num, co2_num)
print(f'Oxy={int(oxy_num, 2)} CO2={int(co2_num, 2)}')
print(f'MUL={int(oxy_num, 2)*int(co2_num, 2)}')

File: 2021/3/puzz2.hs
import System.IO
import Data.Char (digitToInt)
import Data.Typeable (typeOf)
import Data.List.Split (endBy)
import Data.List (foldl')

main :: IO()
main = do
    f <- openFile "input" ReadMode
    contents <- hGetContents f
    let nums = endBy "\n" contents 
    let oxy = extractNum nums mostFilter 0
    let co2 = extractNum nums leastFilter 0
    print $ "Oxy=" ++ show oxy
    print $ "CO2=" ++ show co2
    print $ "MUL=" ++ show (oxy * co2)
    hClose f

type FilterFn = [String] -> Int -> Int -> [String]

extractNum :: [String] -> FilterFn -> Int -> Int
extractNum [a] filter i = toDec a
extractNum nums filter i = extractNum (filter nums ones i) filter (i + 1)
   where ones = countOnes (map (\x -> x !! i) nums)

mostFilter :: FilterFn
mostFilter nums ones i = filter (\x -> (x!!i) == (if ones*2 >= length nums then '1' else '0')) nums

leastFilter :: FilterFn
leastFilter nums ones i = filter (\x -> (x!!i) == (if ones*2 < length nums then '1' else '0')) nums

countOnes :: String -> Int
countOnes = foldr (\x acc -> acc + if (x == '1') then 1 else 0) 0

toDec :: String -> Int
toDec = foldl' (\acc x -> acc * 2 + digitToInt x) 0
